File: src/Scanner.Core/Checks/HiddenNavigationCheck.cs
using System.Text.RegularExpressions;
using Scanner.Core.Rules;

namespace Scanner.Core.Checks;

/// <summary>
/// Checks navigation elements for hidden states that remove them from assistive technology.
/// </summary>
public sealed class HiddenNavigationCheck : ICheck
{
    private static readonly Regex NavRegex = new("<nav(?<attrs>[^>]*)>", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    /// <inheritdoc />
    public string Id => "hidden-navigation";

    /// <inheritdoc />
    public IReadOnlyCollection<string> ApplicableKinds { get; } = new[] { "html", "htm", "cshtml", "razor" };

    /// <inheritdoc />
    public IEnumerable<Issue> Run(CheckContext context, RuleDefinition rule)
    {
        foreach (Match match in NavRegex.Matches(context.Content))
        {
            var attrs = match.Groups["attrs"].Value;
            var ariaHidden = AttributeParser.GetAttributeValue(attrs, "aria-hidden");
            var style = AttributeParser.GetAttributeValue(attrs, "style") ?? string.Empty;
            var hasHidden = TextUtilities.ContainsAttribute(attrs, "hidden");

            if (string.Equals(ariaHidden, "true", StringComparison.OrdinalIgnoreCase)
                || style.Contains("display:none", StringComparison.OrdinalIgnoreCase)
                || style.Contains("visibility:hidden", StringComparison.OrdinalIgnoreCase)
                || hasHidden)
            {
                var line = TextUtilities.GetLineNumber(context.Content, match.Index);
                yield return new Issue(rule.Id, Id, context.FilePath, line, "Navigation element is hidden from assistive tech.", match.Value);
            }
        }
    }
}

File: node_engine/src/checks/HiddenNavigationCheck.js
const { getAttributeValue } = require("./AttributeParser");
const { getLineNumber, containsAttribute } = require("./TextUtilities");

const navRegex = /<nav(?<attrs>[^>]*)>/gi;
const styleWhitespaceRegex = /\s+/g;

const hasHiddenStyle = (style) => {
  if (!style || !style.trim()) {
    return false;
  }

  const normalized = style.replace(styleWhitespaceRegex, "");
  return normalized.toLowerCase().includes("display:none") || normalized.toLowerCase().includes("visibility:hidden");
};

const HiddenNavigationCheck = {
  id: "hidden-navigation",
  applicableKinds: ["html", "htm", "cshtml", "razor"],
  run(context, rule) {
    const issues = [];
    for (const match of context.content.matchAll(navRegex)) {
      const attrs = match.groups.attrs;
      const ariaHidden = getAttributeValue(attrs, "aria-hidden");
      const style = getAttributeValue(attrs, "style") ?? "";
      const hasHidden = containsAttribute(attrs, "hidden", true);

      if ((ariaHidden && ariaHidden.toLowerCase() === "true") || hasHiddenStyle(style) || hasHidden) {
        const line = getLineNumber(context.content, match.index);
        issues.push({
          ruleId: rule.id,
          checkId: HiddenNavigationCheck.id,
          filePath: context.filePath,
          line,
          message: "Navigation element is hidden from assistive tech.",
          evidence: match[0]
        });
      }
    }

    return issues;
  }
};

module.exports = { HiddenNavigationCheck, hasHiddenStyle };
